### PR TITLE
Add test for assembly versions in our packages

### PR DESCRIPTION
**Pull request by wtgodbe**
Part of https://github.com/dotnet/aspnetcore/issues/31074

### CODE REVIEW

File: src/Framework/test/PackageTests.cs
Comment (Pilchie) [inline src/Framework/test/PackageTests.cs:53]: This probably also needs to be removed ;)

File: Directory.Build.props
Comment (dougbu) [reply Directory.Build.props:234]: Also, the backslashes won't work when testing on non-Windows. Use
```suggestion
    <!-- This property points to a folder which includes the (unzipped) contents of our packages, used for testing -->
    <PackagesLayoutRoot>$([MSBuild]::NormalizeDirectory('$(DOTNET_INSTALL_DIR)', '$(ArtifactsObjDir)', 'Packages.Layout', '$(Configuration)'))</PackagesLayoutRoot>
```

File: src/Framework/test/PackageTests.cs
Comment (wtgodbe) [reply src/Framework/test/PackageTests.cs:29]: Didn't see any downside to allowing this extension not to be all lowercase, even if AFAIK it'll never happen

File: src/Framework/test/PackageTests.cs
Comment (wtgodbe) [reply src/Framework/test/PackageTests.cs:75]: Alright, looks like we currently have the following `tools` assemblies in packages:

- `dotnet-sql-cache` has `dotnet-sql-cache.dll`, plus the entire set of .dll's needed to run the tool
- `Microsoft.dotnet-openapi` has `dotnet-openapi.dll`, plus `Microsoft.Build.Locator.dll` in the same folder
- `Microsoft.AspNetCore.Components.WebAssembly.Server` carries a tool folder called `BlazorDebugProxy`, which has `BrowserDebugHost.dll`, `BrowserDebugProxy.dll` (note the difference between BlazorDebugProxy and BrowserDebugProxy), and a few other .dll's needed to run the tool
- `Microsoft.AspNetCore.Components.WebAssembly.Server` carries a copy of the `BlazorDebugProxy` tool, along with a `blazor-devserver.dll` and a matching .exe, plus a large set of .dll's needed to run the tool.

So, I don't think there's an easy rule to apply to all `tools` folders that would cover everything. We could special case these 4, but that's not very future-proof. Any thoughts?

Comment (dougbu) [reply src/Framework/test/PackageTests.cs:75]: Would it work to ignore all assemblies in `tools` that are named `System...` or `Microsoft.*`❔ Allow lists are better than deny lists but this might do the trick

File: src/Framework/test/PackageTests.cs
Comment (wtgodbe) [inline src/Framework/test/PackageTests.cs:109]: This is ugly, @dougbu do you know if there's a cleaner way w/ LINQ to return everything from GetFiles where FileNameWithoutExtension isn't in the below list of things?

Comment (dougbu) [reply src/Framework/test/PackageTests.cs:109]: One idea would be to `Split(...)` the name using `.` and `_` then check whether the first segment is in `{ "aspnetcorev2", "Azure", "Microsoft", "Newtonsoft", "System" }`. I'm sure @davidfowl or @halter73 would have a non-allocating suggestion but this is test code 😀

**Comment by wtgodbe** [comment]
These are the packages that are failing, they wind up in the layout because they're copied there for Helix: https://github.com/dotnet/aspnetcore/blob/b806844373440a538192181652d466b621a14494/eng/helix/helix.proj#L61-L63. Will just skip them

File: src/Framework/test/PackageTests.cs
Comment (dougbu) [inline src/Framework/test/PackageTests.cs:99]: nit: No need to evaluate this each time through the `assembly` list. Suggest something like `var isNetFx = IsNetFx(new DirectoryInfo(tfmDir).Name);`. Could also pass `tfmDir` into `IsNetFx(...)` and create the `DirectoryInfo` instance there.

File: src/Framework/test/PackageTests.cs
Comment (dougbu) [inline src/Framework/test/PackageTests.cs:128]: nit: This is a bit repetitive. Suggest passing `assembly` and an `expectedVersion` (that's already handled the .NET Framework choice) into a method.

Separately, we don't ship any .NET Framework tool assemblies❔

**Review by dougbu — COMMENTED** [review]
A couple of conditions I don't understand but mostly nits for readability

File: src/Framework/test/PackageTests.cs
Comment (wtgodbe) [reply src/Framework/test/PackageTests.cs:66]: https://github.com/dotnet/aspnetcore/blob/b806844373440a538192181652d466b621a14494/eng/helix/helix.proj#L61-L63

File: src/Framework/test/PackageTests.cs
Comment (wtgodbe) [reply src/Framework/test/PackageTests.cs:128]: We don't ship any netfx tool assemblies today. 

We check assemblyVersions the same way in SharedFxTests & TargetingPackTests, I don't feel like fixing all 3 in this PR - maybe as a follow up

**Comment by wtgodbe** [comment]
(Finally) updated for feedback

File: src/Framework/test/PackageTests.cs
Comment (dougbu) [reply src/Framework/test/PackageTests.cs:132]: Sounds fine.

File: src/Framework/test/PackageTests.cs
Comment (dougbu) [inline src/Framework/test/PackageTests.cs:26]: Won't the tests work just fine if run on build agents too❔ I doubt we'll need to stop them running on Helix but suggest
```suggestion
        if (!TestData.VerifyPackageAssemblyVersions() && !SkipOnCIAttribute.OnCI())
```

File: src/Framework/test/PackageTests.cs
Comment (dougbu) [inline src/Framework/test/PackageTests.cs:60]: The above comment doesn't need to be extended, given the information is visible not too far above. However, suggest
```suggestion
        if (!TestData.VerifyPackageAssemblyVersions() && !SkipOnCIAttribute.OnCI())
```

File: src/Framework/test/PackageTests.cs
Comment (dougbu) [inline src/Framework/test/PackageTests.cs:73]: nit
```suggestion
            // Don't test the Shared Framework or Ref pack; assembly versions in those packages are checked elsewhere.
```

File: src/Framework/test/Microsoft.AspNetCore.App.UnitTests.csproj
Comment (dougbu) [inline src/Framework/test/Microsoft.AspNetCore.App.UnitTests.csproj:10]: How does this work when both `$(DOTNET_ROOT)` and `$(ArtifactsObjDir)` are absolute paths❔ I also don't recommend using `$(DOTNET_ROOT)` for layouts because it'll be readonly in some Docker scenarios.
```suggestion
    <PackagesLayoutRoot>$([MSBuild]::NormalizeDirectory('$(ArtifactsObjDir)', 'Packages.Layout', '$(Configuration)'))</PackagesLayoutRoot>
```

File: src/Framework/test/PackageTests.cs
Comment (dougbu) [inline src/Framework/test/PackageTests.cs:34]: Again `$(DOTNET_ROOT)` can be located on a readonly file system
```suggestion
                Environment.GetEnvironmentVariable("HELIX_WORKITEM_ROOT"),
```

File: src/Framework/test/PackageTests.cs
Comment (dougbu) [inline src/Framework/test/PackageTests.cs:49]: Should we mention somewhere that these tests will definitely fail if dev has built for another target `{major}.{minor}` version when run locally❔

File: src/Framework/test/PackageTests.cs
Comment (dougbu) [reply src/Framework/test/PackageTests.cs:132]: Did you file a follow up (nit) issue❔

**Review by dougbu — COMMENTED** [review]
Almost there 👍 Mostly nits but a few things should change before merge&hellip;

File: src/Framework/test/PackageTests.cs
Comment (wtgodbe) [reply src/Framework/test/PackageTests.cs:134]: I don't think this is important enough to have an issue for (or even to proactively fix, if I'm being honest)

**Comment by wtgodbe** [comment]
Should be all good now, assuming CI passes.

File: src/Framework/test/PackageTests.cs
Comment (wtgodbe) [reply src/Framework/test/PackageTests.cs:51]: Added a comment

File: src/Framework/test/PackageTests.cs
Comment (dougbu) [inline src/Framework/test/PackageTests.cs:25]: Nit because the restriction here is about the CI. But it's not imperative that the comment is perfect.
```suggestion
        // Normally, should only run this test in official builds. The test will work locally so long as you're building w/
```

Something similar for `PackageAssembliesHaveExpectedAssemblyVersions()`.

**Review by dougbu — APPROVED** [review]

**Pull request merged**
